femtools/Detector_Tools.F90

```diff
@@ -612,24 +633,102 @@ subroutine set_detector_coords_from_python(values, ndete, func, time)
 
   end subroutine set_detector_coords_from_python
 
-  subroutine set_particle_attribute_from_python(attributes, positions, nparts, func, time, dt)
-    !!< Given the particle position and time, evaluate the specified python function for a group of particles
-    !!< specified in the string func at that location. 
-    real, dimension(:), intent(inout) :: attributes
+  !> Evaluate a set of fields on particles
+  subroutine evaluate_particle_fields(npart, state, ele, lcoords, names, phases, counts, vals, dim)
+    !! Number of particles
+    integer, intent(in) :: npart
+    !! Model state structure
+    type(state_type), dimension(:), intent(in) :: state
+    !! Elements containing particles
+    integer, dimension(:), intent(in) :: ele
+    !! Local coordinates of particles in elements
+    real, dimension(:,:), intent(in) :: lcoords
+    !! Names of fields to be evaluated
+    type(attr_names_type), intent(in) :: names
+    !! Phases of each field
+    type(field_phase_type), intent(in) :: phases
+    !! Scalar/vector/tensor counts of each field
+    integer, dimension(3), intent(in) :: counts
+    !! Output array
+    real, dimension(:,:), intent(out) :: vals
+    !! Geometric dimension
+    integer, intent(in) :: dim
+
+    integer :: i, j
+    integer :: field_idx, phase
+    type(scalar_field), pointer :: sfield
+    type(vector_field), pointer :: vfield
+    type(tensor_field), pointer :: tfield
+
+    field_idx = 1
+
+    ! scalar fields
+    do i = 1, counts(1)
+      phase = phases%s(i)
+      sfield => extract_scalar_field(state(phase), names%s(i))
+
+      do j = 1, npart
+        vals(field_idx,j) = eval_field(ele(j), sfield, lcoords(:,j))
+      end do
+
+      field_idx = field_idx + 1
+    end do
+
+    ! vector fields
+    do i = 1, counts(2)
+      phase = phases%v(i)
+      vfield => extract_vector_field(state(phase), names%v(i))
+
+      do j = 1, npart
+        vals(field_idx:field_idx+dim-1,j) = eval_field(ele(j), vfield, lcoords(:,j))
+      end do
+
+      field_idx = field_idx + dim
+    end do
+
+    ! tensor fields
+    do i = 1, counts(3)
+      phase = phases%t(i)
+      tfield => extract_tensor_field(state(phase), names%t(i))
+
+      do j = 1, npart
+        vals(field_idx:field_idx+dim**2-1,j) = reshape( &
+             eval_field(ele(j), tfield, lcoords(:,j)), &
+             [ dim**2 ])
+      end do
+
+      field_idx = field_idx + dim**2
+    end do
+  end subroutine evaluate_particle_fields
+
+  !> Given the particle position and time, evaluate the specified python function for a group of particles
+  !! specified in the string func at that location.
+  subroutine set_particle_scalar_attribute_from_python(attributes, positions, nparts, natt, func, time, dt, is_array)
+    real, dimension(:,:), intent(out) :: attributes
```

Comment:
> Is this `natt X nparts`
Yes -- I've clarified that

> Since this is the fortran interface, do we need to provide nparts and natt seperately? Can we note use size(attributes,1) and size(attributes,2)?

Sure, that's a good idea.

Reply:
Actually looking through the following vector and tensor functions, where `natt` is different from `size(attributes,1)`, it would be more consistent to still specify this. We can still drop `nparts` though.